src/libraries/DecimalNormalization.sol

```diff
@@ -0,0 +1,109 @@
+            "DecimalNormalization: amount decimals must be less than or equal to token decimals"
+        );
+
+        // Scale the amount to match token decimals
+        return scaleDecimals(scaledAmount, tokenDecimals - parsedDecimals);
+    }
+
+    /// @notice Validates that the amount string has a valid format
+    /// @param components The split components of the amount string
+    /// reverts if components are invalid
+    function validateAmountFormat(string[] memory components) internal pure {
+        // Check for invalid format (more than one decimal point)
+        require(components.length <= 2, "DecimalNormalization: invalid amount");
+
+        // If there's a decimal part, ensure it's not empty
+        if (components.length == 2) {
+            require(bytes(components[1]).length != 0, "DecimalNormalization: decimals cannot be 0");
+        }
+    }
+
+    /// @notice Checks if the amount represents a non-zero value
+    /// @param wholePart The whole number part of the amount
+    /// @param decimalPart The decimal part of the amount
+    /// @return True if either part results in a non-zero value
+    function isNonZeroAmount(uint256 wholePart, uint256 decimalPart) internal pure returns (bool) {
+        return wholePart > 0 || decimalPart > 0;
+    }
+
+    /// @notice Parses a string amount into its components
+    /// @param amount The amount string to parse
+    /// @return components The parsed amount components
+    function parseAmountComponents(string memory amount) internal pure returns (AmountComponents memory components) {
+        string[] memory stringComponents = amount.split(".");
+        validateAmountFormat(stringComponents);
+
+        components.integer = vm.parseUint(stringComponents[0]);
+        components.hasDecimals = stringComponents.length == 2;
+
+        if (components.hasDecimals) {
+            components.decimal = vm.parseUint(stringComponents[1]);
+            components.decimalPlaces = uint8(bytes(stringComponents[1]).length);
+            require(components.decimalPlaces <= 18, "DecimalNormalization: decimals must be less than or equal to 18");
+        }
+
+        // Ensure the final amount is non-zero
+        require(
+            isNonZeroAmount(components.integer, components.decimal), "DecimalNormalization: amount must be non-zero"
+        );
+
+        return components;
+    }
+
+    /// @notice returns the amount with the length of the decimals parsed
+    /// @param amount The amount to parse
+    /// @return The amount and the number of decimals
+    function parseDecimals(string memory amount) internal pure returns (uint256, uint8) {
+        AmountComponents memory components = parseAmountComponents(amount);
+
+        if (!components.hasDecimals || components.decimal == 0) {
+            return (components.integer, 0);
+        }
+
+        uint256 finalAmount = components.integer * (10 ** uint256(components.decimalPlaces)) + components.decimal;
```

Comment:
You shouldn't be scaling amounts in a `parseDecimals` function, that's confusing.

Reply:
I would probably have two functions:
- `parseAmountComponents`: Gets a string amount and a decimal count and returns an AmountComponents struct.
- `normalizeTokenAmount`: Gets a string amount and a token, does some verification, obtains the decimals, calls `parseAmountComponents`, scales up the integer part, and returns the scaled up integer part plus the decimal part.

Reply:
Wrote a PR instead of going back and forth:
https://github.com/solidity-labs-io/superchain-ops/pull/36